transports/webrtc-websys/src/upgrade.rs

```diff
@@ -0,0 +1,56 @@
+use super::Error;
+use crate::connection::RtcPeerConnection;
+use crate::sdp;
+use crate::Connection;
+use libp2p_identity::{Keypair, PeerId};
+use libp2p_webrtc_utils::noise;
+use libp2p_webrtc_utils::Fingerprint;
+use send_wrapper::SendWrapper;
+use std::net::SocketAddr;
+
+/// Upgrades an outbound WebRTC connection by creating the data channel
+/// and conducting a Noise handshake
+pub(crate) async fn outbound(
+    sock_addr: SocketAddr,
+    remote_fingerprint: Fingerprint,
+    id_keys: Keypair,
+) -> Result<(PeerId, Connection), Error> {
+    let fut = SendWrapper::new(outbound_inner(sock_addr, remote_fingerprint, id_keys));
+    fut.await
+}
+
+/// Inner outbound function that is wrapped in [SendWrapper]
+async fn outbound_inner(
+    sock_addr: SocketAddr,
+    remote_fingerprint: Fingerprint,
+    id_keys: Keypair,
+) -> Result<(PeerId, Connection), Error> {
+    let rtc_peer_connection = RtcPeerConnection::new(remote_fingerprint.algorithm()).await?;
+
+    // Create stream for Noise handshake
+    // Must create data channel before Offer is created for it to be included in the SDP
+    let (channel, listener) = rtc_peer_connection.new_handshake_stream();
+    drop(listener);
+
+    let ufrag = libp2p_webrtc_utils::sdp::random_ufrag();
+
+    let offer = rtc_peer_connection.create_offer().await?;
+    let munged_offer = sdp::offer(offer, &ufrag);
+    rtc_peer_connection
+        .set_local_description(munged_offer)
+        .await?;
+
+    let answer = sdp::answer(sock_addr, remote_fingerprint, &ufrag);
+    rtc_peer_connection.set_remote_description(answer).await?;
+
+    let local_fingerprint = rtc_peer_connection.local_fingerprint()?;
```

Comment:
The `browser-webrtc` example fails to connect to the WebRTC endpoint on `localhost` **when using Firefox**. It works fine on Chromium.

I first assumed that this is due to Firefox not supporting `local_fingerprint`. See https://github.com/libp2p/js-libp2p-webrtc/pull/109 and https://github.com/libp2p/js-libp2p-webrtc/issues/96 for a similar issue on the JS side.

I remember Firefox not being able to connect to `localhost` on WebRTC. Is that the reason @DougAnderson444?

Reply:
Since we are building the fingerprint in our functions, I suspect it is indeed the `localhost` issue @tomaka mentioned earlier that would cause this headache with Firefox. I don't think the `if firefox && target == "localhost" { return Err }` has been injected anywhere yet.

Reply:
Ah that is a bummer. We should definitely add said line with a link to the bug report to avoid headaches for users.

In regards to the fingerprint, should we merge without support for FF or add it in? As far as I remember, we'd need to parse the fingerprint from an SDP right?

Reply:
> In regards to the fingerprint, should we merge without support for FF or add it in? As far as I remember, we'd need to parse the fingerprint from an SDP right?

Unless I am missing something, that is already the case. See `connection.rs` `parse_fingerprint`:

https://github.com/libp2p/rust-libp2p/pull/4248/files#diff-33a723dd57c111e6156b9a2a1bee557456217322a2478302bf6c19859ad4ef7aR276-R293

@DougAnderson444 please correct me if I am wrong.

> I don't think the `if firefox && target == "localhost" { return Err }` has been injected anywhere yet.

@DougAnderson444 mind adding this? I am sure many folks will be grateful for it!

Reply:
> Unless I am missing something, that is already the case.

Yes our code parses the fingerprint, so there is no need to rely on Firefox to do this.

> mind adding this?

I am writing a function to test the user agent and window location and return a Transport error if it's local firefox

Reply:
All done